### ISSUE TITLE
<replacementProperty> fails to work with an empty <value> tag

### ISSUE THREAD

**Issue by toolforger** [git-commit-id/maven-git-commit-id-plugin#389]
### Describe the bug
Tried to eliminate the `-SNAPSHOT` suffix from `git.build.version`, using an empty `<value>` tag.
The plugin would not apply the rule in that case (which does not make any sense).

### Steps to Reproduce; Expected and Observed Behaviour

Given a tag of "10.0-SNAPSHOT", this produces `git.build.version.nosnapshot=10.1` as intended:

```xml
           <plugin>
                <groupId>pl.project13.maven</groupId>
                <artifactId>git-commit-id-plugin</artifactId>
                <version>2.2.5</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>revision</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <replacementProperties>
                        <replacementProperty>
                            <property>git.build.version</property>
                            <token>(.)-SNAPSHOT$</token> <!-- workaround: detect any character -->
                            <value>$1</value> <!-- workaround: have a nonempty <value> tag -->
                            <propertyOutputSuffix>nosnapshot</propertyOutputSuffix>
                        </replacementProperty>
                    </replacementProperties>
                </configuration>
            </plugin>
```

If I use

```xml
                            <token>-SNAPSHOT$</token>
                            <value></value>
```

or

```xml
                            <token>-SNAPSHOT$</token>
                            <value/>
```

I get `git.build.version.nosnapshot=10.1-SNAPSHOT`, i.e. the replacement is not happening.

No error messages or stack traces.

### Additional context
For reproducibility please provide the following:

JDK Version: 1.8.0_152
Maven Version: 3.3.9
OS Info: Arch: amd64 Family: windows Name: windows 7 Version: 6.1
Running as a Maven build inside IntelliJ 2018.2; command line is `C:\Daten\software\jdk-8u152-windows-x64\bin\java.exe -Dmaven.multiModuleProjectDirectory=C:\develop\id-gui -Dmaven.home=C:\Users\tkd6u\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\182.4505.22\plugins\maven\lib\maven3 -Dclassworlds.conf=C:\Users\tkd6u\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\182.4505.22\plugins\maven\lib\maven3\bin\m2.conf -javaagent:C:\Users\tkd6u\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\182.4505.22\lib\idea_rt.jar=64767:C:\Users\tkd6u\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\182.4505.22\bin -Dfile.encoding=UTF-8 -classpath C:\Users\tkd6u\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\182.4505.22\plugins\maven\lib\maven3\boot\plexus-classworlds-2.5.2.jar org.codehaus.classworlds.Launcher -Didea.version=2018.2.4 --errors --update-snapshots compile`

**Comment by toolforger** [comment]
Note that the workaround above won't work if a property is to be completely replaced with an empty string.
Fingers crossed - I hope I won't need an empty property.

**Comment by TheSnoozer** [comment]
Hi, thanks for reporting your issue/problem here.
Unfortunately I fear that's maven's fault and a similar issue as outlined in https://github.com/ktoso/maven-git-commit-id-plugin/issues/374.

The root issue I think is that maven just doesn't manage to tell the plugin the right settings (e.g. in that case the plugin doesn't even see that you have set the property to something empty).

Related Maven issues:
- https://issues.apache.org/jira/browse/MNG-6434
- https://issues.apache.org/jira/browse/MNG-5725

In fact the only thing i can do is actually suggest a similar workaround (https://github.com/ktoso/maven-git-commit-id-plugin/issues/374#issuecomment-411850735) and thayou already seem to have in place:
```t 
<replacementProperties>
                <replacementProperty>
                            <property>git.build.version</property>
                            <token>(.*)(-SNAPSHOT)?$</token>
                            <value>$1</value>
                            <propertyOutputSuffix>nosnapshot</propertyOutputSuffix>
                </replacementProperty>
</replacementProperties>
```

**Comment by toolforger** [comment]
Well, the workaround didn't work out in the end, so this is still unsolved for me.
What about making `<value>` an empty string by default? Then specifying it empty or not at all doesn't make a difference. Sure it can create surprising behaviour, but well this is Maven anyway.

**Comment by TheSnoozer** [comment]
Right,
I should have tested the regex :')
In all fairness there are plenty of way to remove the optional ending string, its just a matter of finding the "right" regex.

See here for some inspiration how to get this working:
- https://regex101.com/r/bA2Tlx/1/
- https://regex101.com/r/bPFBwi/1/
- https://regex101.com/r/wTJXhO/1/

You can for example specify the regex with CDATA-Tags:
```xml
<token><![CDATA[^([^-SNAPSHOT]*)(.*)$]]></token>
```

**Comment by toolforger** [comment]
This issue is not about the regex.
It is about '<value>': You cannot specify an empty replacement string.

**Comment by TheSnoozer** [comment]
Yeah I understand that. My suggested workaround is to build a regex that matches everything besides ```-SNAPSHOT``` that will also make use of regex groups to then eliminate the ```-SNAPSHOT``` (optional) suffix from ```git.build.version```. Essentially use the power of regex to create a substring.

```xml
<replacementProperties>
                <replacementProperty>
                            <property>git.build.version</property>
                            <token><![CDATA[^([^-SNAPSHOT]*)(.*)$]]></token>
                            <value>$1</value>
                            <propertyOutputSuffix>nosnapshot</propertyOutputSuffix>
                </replacementProperty>
</replacementProperties>
```

Seems to work perfectly fine.

**Comment by TheSnoozer** [comment]
Hi thanks again for opening an issue for this.
This has now been addressed by https://github.com/git-commit-id/maven-git-commit-id-plugin/pull/400 and will become available with the next 3.0 release.

**Comment by toolforger** [comment]
Thanks :-)
